test/test_browser.py

```diff
@@ -4851,7 +4853,7 @@ def test_single_file_locate_file(self):
   # Tests that SINGLE_FILE works as intended in a Worker in JS output
   def test_single_file_worker_js(self):
     self.compile_btest('browser_test_hello_world.c', ['-o', 'test.js', '--proxy-to-worker', '-sSINGLE_FILE'])
-    create_file('test.html', '<script src="test.js"></script>')
+    create_file('test.html', '<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"></head><body><script src="test.js"></script></body></html>')
```

Comment:
Does this test break without this change?

Does the charset in the `<head>` somehow effect the script tag?

Does the mean that some/all users of `SINGLE_FILE` would ned to add an explicit charset to their html?

Reply:
Our existing shell files have [already ~from the dawn of time](https://github.com/emscripten-core/emscripten/commit/0ea9684acb5206669788bee5ed557a070f9bfe4e#diff-3dfaf4a1998bb76cd59e6121ae71c10987c32fc478918aa92c63bdb1d61d1c06R4) had a `<meta charset='utf-8'>` directive. That is part of the general "best practices" boilerplate.

So we can expect quite safely that all the custom shell files that users have monkeyed off of any of the existing shell files we provide, will be UTF-8 (unless they have explicitly removed the meta charset directive, which is a bad idea in general)

About [98.2% of all websites utilize UTF-8](https://w3techs.com/technologies/cross/character_encoding/ranking), it is practically the only encoding used on the web. I don't really know of anyone who would use any other encoding on their web pages anymore.

It is just that our tests scripts have omitted the meta charset='utf-8' directive.

(If users did not want for some reason to specify that meta charset directive, they can also save the HTML file with a [UTF-8 BOM marker](https://en.wikipedia.org/wiki/Byte_order_mark#UTF-8), or they can send a `Content-Type: application/html; charset=utf-8` HTTP header, but typically the meta charset directive is the simplest way)

Reply:
> To be clear, without this change, this test is broken?

That's correct, the test would not work otherwise.

> I wonder if we should write a test to verify that `-sSINGLE_FILE` output is broken-by-default without `UTF-8` encoding but that adding `<meta charset='utf-8'>` or disabling `SINGLE_FILE_BINARY_ENCODE` fixes it?

I am not sure what would be the benefit of such test?

If we wanted to go the extra mile, I think we would want to either generate a UTF-8 BOM into the HTML file, or issue a link time warning if generated HTML file contains UTF-8 code points, but the HTML file does not contain a `<meta charset='utf-8'>` directive.

> I guess we don't have any other non-ascii strings in our generated code otherwise `<meta charset='utf-8'>` would be required regardless of `SINGLE_FILE` and / or `SINGLE_FILE_BINARY_ENCODE`?

That is true. We (Emscripten) don't, but if users would have any UTF-8 chars they contain in JS libraries, then that would require them to specify the charset.

Reply:
> issue a link time warning if generated HTML file contains UTF-8 code points, but the HTML file does not contain a `<meta charset='utf-8'>` directive.

This would a useful warning for users who generate HTML output but my understanding is that the vast majority of our users generate JavaScript and then write their own HTML separetely.  I'm curious, does unity use emscripten to generate HTML?  Is so I guess that would be big use case for HTML output in production.   Otherwise I don't know of any.